processing/src/main/java/org/apache/druid/segment/column/CapabilitiesBasedFormat.java

```diff
@@ -0,0 +1,151 @@
+  {
+    this.capabilities = capabilities;
+  }
+
+  @Override
+  public DimensionHandler getColumnHandler(String columnName)
+  {
+    return DimensionHandlerUtils.getHandlerFromCapabilities(columnName, capabilities, null);
+  }
+
+  @Override
+  public DimensionSchema getColumnSchema(String columnName)
+  {
+    return getColumnHandler(columnName).getDimensionSchema(capabilities);
+  }
+
+  @Override
+  public ColumnFormat merge(@Nullable ColumnFormat otherFormat)
+  {
+    if (otherFormat == null) {
+      return this;
+    }
+
+    ColumnCapabilitiesImpl merged = ColumnCapabilitiesImpl.copyOf(this.toColumnCapabilities());
+    ColumnCapabilitiesImpl otherSnapshot = ColumnCapabilitiesImpl.copyOf(otherFormat.toColumnCapabilities());
+
+    if (!Objects.equals(merged.getType(), otherSnapshot.getType())
+        || !Objects.equals(merged.getElementType(), otherSnapshot.getElementType())) {
+      final String mergedType = merged.getType() == null ? null : merged.asTypeString();
+      final String otherType = otherSnapshot.getType() == null ? null : otherSnapshot.asTypeString();
+      throw new ISE(
+          "Cannot merge columns of type[%s] and [%s]",
+          mergedType,
+          otherType
+      );
+    } else if (!Objects.equals(merged.getComplexTypeName(), otherSnapshot.getComplexTypeName())) {
+      throw new ISE(
+          "Cannot merge columns of type[%s] and [%s]",
+          merged.getComplexTypeName(),
+          otherSnapshot.getComplexTypeName()
+      );
+    }
+
+    merged.setDictionaryEncoded(merged.isDictionaryEncoded().or(otherSnapshot.isDictionaryEncoded()).isTrue());
+    merged.setHasMultipleValues(merged.hasMultipleValues().or(otherSnapshot.hasMultipleValues()).isTrue());
+    merged.setDictionaryValuesSorted(
+        merged.areDictionaryValuesSorted().and(otherSnapshot.areDictionaryValuesSorted()).isTrue()
+    );
+    merged.setDictionaryValuesUnique(
+        merged.areDictionaryValuesUnique().and(otherSnapshot.areDictionaryValuesUnique()).isTrue()
+    );
+    merged.setHasNulls(merged.hasNulls().or(otherSnapshot.hasNulls()).isTrue());
+    // When merging persisted queryableIndexes in the same ingestion job,
+    // all queryableIndexes should have the exact same hasBitmapIndexes flag set which is set in the ingestionSpec.
+    // One exception is null-only columns as they always do NOT have bitmap indexes no matter whether the flag is set
+    // in the ingestionSpec. As a result, the mismatch checked in the if clause below can happen
+    // when one of the columnCapability is from a real column and another is from a null-only column.
+    // See NullColumnPartSerde for how columnCapability is created for null-only columns.
+    // When the mismatch is found, we prefer the flag set in the ingestionSpec over
+    // the columnCapability of null-only columns.
+    if (merged.hasBitmapIndexes() != otherSnapshot.hasBitmapIndexes()) {
+      merged.setHasBitmapIndexes(false);
+    }
```

Comment:
yeah i copied this code from elsewhere, but agree it seems strange, looking into it. the comment isn't even true, null column part serde sets 'has bitmap indexes' to true because it has a 'null' index supplier which allows it to use all true/all false bitmaps if the filter matches null...